qi-lib/flow/core/compiler.rkt

```diff
@@ -485,10 +486,6 @@ the DSL.
                (qi0->racket (~> (-< (~> (gen args) △) _)
                                 onex))))]))
 
-  (define (literal-parser stx)
-    (syntax-parse stx
-      [val:literal #'(qi0->racket (gen val))]))
-
```

Comment:
Commit 0fb1afa says this has been in the expander for a while; I think it was effectively introduced (and then several times refactored) by commit 5e1e06eb as part of the flow macro (this was Oct. 2021 and is in `main`!).

This does make me wonder what machinery besides coverage reports we have to make sure that all passes, parsers, etc., are tied together. For example, with the compiler->expander pipeline, I think there are several stages to trace through to follow an input flow syntax to an output syntax: how do we make the sure each "pass" or "stage" handles all (and only) the outputs of its previous stage?

Reply:
I definitely agree the implementation has reached a level of complexity where having many different and complementary ways to test things would be valuable and would save us a lot of time. I'll summarize what we have today and what is planned -- lmk if this addresses your concern adequately or if you feel more is needed:

- we have regular unit tests validating the semantics of the language (e.g. `qi-test/tests/flow.rkt`). In addition to local tests of individual forms, this includes some nonlocal flows like the counterexamples you found, and in general we'd want these to tell us if we change the meaning of the language (including unsound rules).
- there are starter "expander" tests (`qi-test/tests/expander.rkt`) that verify that surface expressions are correctly expanded to core expressions (but it can only compare the result as datums for now)
- there are compiler "rules" tests (`qi-test/tests/compiler/rules.rkt`) that check individual rewrite rules, and also check transformations effected by overall individual compiler passes like normalization (which may repeatedly apply sets of rules to a fixed point). These verify that when given correct input they would produce the expected optimized output.
- compiler "semantics" tests (`qi-test/tests/compiler/semantics.rkt`) that check various combinations of patterns that are matched by optimization rules, to verify that they produce the same output (including raising errors) as the original expression. These patterns are specific to the compiler implementation and that's why they aren't language level (regular) unit tests.

Planned:
The above tests would validate that different components behave as expected _when_ they are invoked, but we still don't have a way to know _whether_ they are invoked for a particular expression. Some folks on discourse suggested "trace" testing, where we add logging to each pass of the compiler and then verify that a given expression goes through the correct sequence of compiler passes by capturing logs from invoking the compiler. 

Wdyt? Anything else that would be useful?

Reply:
All sounds reasonable. 

One thing I was considering was: how do I check what the syntax (or other input) to each "stage" is? Such as compiler, expander, optimizer, etc. What is the output?

I would call them languages, but HtDP might call them data definitions 😅 basically a "this is what we can expect to see coming in and out, and code should handle all these cases." That would pave way to follow the code through each pass, for example, to trace a specific form through the pipeline.

Reply:
I'll give this some more thought. It would be nice to be able to trace an expression like this on-demand in a testing/REPL context. The macro stepper should be one way to gain visibility of this kind but currently doesn't give us the most support due to some APIs being private and not supported IIUC.